### ISSUE TITLE
onExit handlers of DAG-task (A) never called if there is another DAG-task (B) depending on DAG-task (A)

### ISSUE THREAD

**Issue by arbielsk** [argoproj/argo#3395]
Checklist:

 * [x] I've included the version.
 * [x] I've included reproduction steps.
 * [x] I've included the workflow YAML.
 * [x] I've included the logs.
 
**What happened**:
I created a dag where each task has an exit handler. In the example below, step-3 depends on step-2. The onExit handler of step-2 is never called. However the onExit handlers for step-1 and step-3 are called as expected.

**What you expected to happen**:
I expected onExit handlers to be called on every task, no matter if another task depends on it.

**How to reproduce it (as minimally and precisely as possible)**:
```
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  name: exit-handler-bug-example
spec:
  entrypoint: dag
  templates:
    - name: dag
      dag:
        tasks:
          - name: step-1
            onExit: exit-handler
            template: step-template
          - name: step-2
            onExit: exit-handler
            template: step-template
          - name: step-3
            dependencies: [step-2]
            onExit: exit-handler
            template: step-template
    - name: exit-handler
      dag:
        tasks:
          - name: exit-1
            template: sleep-10
    - name: sleep-10
      container:
        image: alpine:latest
        command:
          - sh
          - -c
        args:
          - "sleep 10 && echo exit-handler-step-{{pod.name}}"
    - name: step-template
      container:
        image: alpine:latest
        command:
          - sh
          - -c
        args:
          - "sleep 5 && echo step {{pod.name}}"
```

**Anything else we need to know?**:

**Environment**:

- Argo version:
```
argo: v2.9.0-rc4
  BuildDate: 2020-06-26T21:55:04Z
  GitCommit: 5b109bcb9257653ecbae46e6315c8d65842de58a
  GitTreeState: clean
  GitTag: v2.9.0-rc4
  GoVersion: go1.13.4
  Compiler: gc
  Platform: darwin/amd64
```
- Kubernetes version :
```
clientVersion:
  buildDate: "2020-01-15T08:26:26Z"
  compiler: gc
  gitCommit: e7f962ba86f4ce7033828210ca3556393c377bcc
  gitTreeState: clean
  gitVersion: v1.16.6-beta.0
  goVersion: go1.13.5
  major: "1"
  minor: 16+
  platform: darwin/amd64
serverVersion:
  buildDate: "2020-01-15T08:18:29Z"
  compiler: gc
  gitCommit: e7f962ba86f4ce7033828210ca3556393c377bcc
  gitTreeState: clean
  gitVersion: v1.16.6-beta.0
  goVersion: go1.13.5
  major: "1"
  minor: 16+
  platform: linux/amd64
```

**Other debugging information (if applicable)**:

- workflow result:

```
DEBU[0000] CLI version                                   version="{v2.9.0-rc4 2020-06-26T21:55:04Z 5b109bcb9257653ecbae46e6315c8d65842de58a v2.9.0-rc4 clean go1.13.4 gc darwin/amd64}"
DEBU[0000] Client options                                opts="{{ false false}  0x2277680 0xc0002e2dc0}"
Name:                exit-handler-bug-example
Namespace:           argo
ServiceAccount:      default
Status:              Succeeded
Conditions:          
 Completed           True
Created:             Tue Jul 07 10:10:16 +0200 (6 minutes ago)
Started:             Tue Jul 07 10:10:16 +0200 (6 minutes ago)
Finished:            Tue Jul 07 10:10:50 +0200 (6 minutes ago)
Duration:            34 seconds
ResourcesDuration:   1m23s*(1 cpu),1m23s*(100Mi memory)

STEP                         TEMPLATE       PODNAME                              DURATION  MESSAGE
 ✔ exit-handler-bug-example  dag                                                             
 ├-✔ step-1                  step-template  exit-handler-bug-example-3071691002  10s         
 ├-✔ step-2                  step-template  exit-handler-bug-example-3054913383  8s          
 ├-✔ step-3                  step-template  exit-handler-bug-example-3038135764  8s          
 ├-✔ step-1.onExit           exit-handler                                                    
 | └-✔ exit-1                sleep-10       exit-handler-bug-example-2218368392  14s         
 └-✔ step-3.onExit           exit-handler                                                    
   └-✔ exit-1                sleep-10       exit-handler-bug-example-1275286746  13s
```

- executor logs:

```
NOT APPLICABLE AS POD IS NEVER STARTED: kubectl logs <failedpodname> -c init
NOT APPLICABLE AS POD IS NEVER STARTED: kubectl logs <failedpodname> -c wait
```

- workflow-controller logs:

```
kubectl logs -n argo $(kubectl get pods -l app=workflow-controller -n argo -o name)
time="2020-07-07T08:10:16Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:16Z" level=info msg="Updated phase  -> Running" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:16Z" level=info msg="DAG node exit-handler-bug-example initialized Running" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:16Z" level=info msg="All of node exit-handler-bug-example.step-1 dependencies [] completed" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:16Z" level=info msg="Pod node exit-handler-bug-example-3071691002 initialized Pending" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:16Z" level=info msg="Created pod: exit-handler-bug-example.step-1 (exit-handler-bug-example-3071691002)" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:16Z" level=info msg="All of node exit-handler-bug-example.step-2 dependencies [] completed" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:16Z" level=info msg="Pod node exit-handler-bug-example-3054913383 initialized Pending" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:16Z" level=info msg="Created pod: exit-handler-bug-example.step-2 (exit-handler-bug-example-3054913383)" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:16Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=1405366 workflow=exit-handler-bug-example
time="2020-07-07T08:10:16Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:16Z" level=info msg="Updating node exit-handler-bug-example-3071691002 message: ContainerCreating"
time="2020-07-07T08:10:16Z" level=info msg="Updating node exit-handler-bug-example-3054913383 message: ContainerCreating"
time="2020-07-07T08:10:17Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=1405373 workflow=exit-handler-bug-example
time="2020-07-07T08:10:17Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:17Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:21Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:21Z" level=info msg="Updating node exit-handler-bug-example-3054913383 status Pending -> Running"
time="2020-07-07T08:10:21Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=1405393 workflow=exit-handler-bug-example
time="2020-07-07T08:10:21Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:21Z" level=info msg="Updating node exit-handler-bug-example-3071691002 status Pending -> Running"
time="2020-07-07T08:10:22Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=1405394 workflow=exit-handler-bug-example
time="2020-07-07T08:10:22Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:22Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:26Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:26Z" level=info msg="Setting node exit-handler-bug-example-3054913383 outputs"
time="2020-07-07T08:10:26Z" level=info msg="Updating node exit-handler-bug-example-3054913383 status Running -> Succeeded"
time="2020-07-07T08:10:26Z" level=info msg="All of node exit-handler-bug-example.step-3 dependencies [step-2] completed" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:26Z" level=info msg="Pod node exit-handler-bug-example-3038135764 initialized Pending" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:26Z" level=info msg="Created pod: exit-handler-bug-example.step-3 (exit-handler-bug-example-3038135764)" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:26Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=1405411 workflow=exit-handler-bug-example
time="2020-07-07T08:10:26Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:26Z" level=info msg="Labeled pod argo/exit-handler-bug-example-3054913383 completed"
time="2020-07-07T08:10:26Z" level=info msg="Setting node exit-handler-bug-example-3071691002 outputs"
time="2020-07-07T08:10:26Z" level=info msg="Updating node exit-handler-bug-example-3038135764 message: ContainerCreating"
time="2020-07-07T08:10:27Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=1405418 workflow=exit-handler-bug-example
time="2020-07-07T08:10:27Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:27Z" level=info msg="Labeled pod argo/exit-handler-bug-example-3054913383 completed"
time="2020-07-07T08:10:27Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:28Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:28Z" level=info msg="Updating node exit-handler-bug-example-3071691002 status Running -> Succeeded"
time="2020-07-07T08:10:28Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:28Z" level=info msg="DAG node exit-handler-bug-example-1359570492 initialized Running" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:28Z" level=info msg="All of node step-1.onExit.exit-1 dependencies [] completed" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:28Z" level=info msg="Pod node exit-handler-bug-example-2218368392 initialized Pending" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:28Z" level=info msg="Created pod: step-1.onExit.exit-1 (exit-handler-bug-example-2218368392)" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:29Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=1405430 workflow=exit-handler-bug-example
time="2020-07-07T08:10:29Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:29Z" level=info msg="Labeled pod argo/exit-handler-bug-example-3071691002 completed"
time="2020-07-07T08:10:29Z" level=info msg="Updating node exit-handler-bug-example-2218368392 message: ContainerCreating"
time="2020-07-07T08:10:29Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:29Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=1405440 workflow=exit-handler-bug-example
time="2020-07-07T08:10:29Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:29Z" level=info msg="Labeled pod argo/exit-handler-bug-example-3071691002 completed"
time="2020-07-07T08:10:29Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:29Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:29Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:31Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:31Z" level=info msg="Updating node exit-handler-bug-example-3038135764 status Pending -> Running"
time="2020-07-07T08:10:31Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:31Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=1405447 workflow=exit-handler-bug-example
time="2020-07-07T08:10:31Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:31Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:33Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:33Z" level=info msg="Updating node exit-handler-bug-example-2218368392 status Pending -> Running"
time="2020-07-07T08:10:33Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:34Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=1405453 workflow=exit-handler-bug-example
time="2020-07-07T08:10:34Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:34Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:36Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:36Z" level=info msg="Setting node exit-handler-bug-example-3038135764 outputs"
time="2020-07-07T08:10:36Z" level=info msg="Updating node exit-handler-bug-example-3038135764 status Running -> Succeeded"
time="2020-07-07T08:10:36Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:36Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:36Z" level=info msg="DAG node exit-handler-bug-example-2933386262 initialized Running" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:36Z" level=info msg="All of node step-3.onExit.exit-1 dependencies [] completed" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:36Z" level=info msg="Pod node exit-handler-bug-example-1275286746 initialized Pending" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:36Z" level=info msg="Created pod: step-3.onExit.exit-1 (exit-handler-bug-example-1275286746)" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:36Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=1405466 workflow=exit-handler-bug-example
time="2020-07-07T08:10:36Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:36Z" level=info msg="Labeled pod argo/exit-handler-bug-example-3038135764 completed"
time="2020-07-07T08:10:36Z" level=info msg="Updating node exit-handler-bug-example-1275286746 message: ContainerCreating"
time="2020-07-07T08:10:36Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:36Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:37Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=1405474 workflow=exit-handler-bug-example
time="2020-07-07T08:10:37Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:37Z" level=info msg="Labeled pod argo/exit-handler-bug-example-3038135764 completed"
time="2020-07-07T08:10:37Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:37Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:37Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:37Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:37Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:37Z" level=info msg="ib-vip-sp21-21-07-390-9-2006-1-0-m7hk7 is suspended, skipping execution" namespace=argo workflow=ib-vip-sp21-21-07-390-9-2006-1-0-m7hk7
time="2020-07-07T08:10:38Z" level=info msg="ib-vip-sp21-21-07-390-9-2006-1-0-htf86 is suspended, skipping execution" namespace=argo workflow=ib-vip-sp21-21-07-390-9-2006-1-0-htf86
time="2020-07-07T08:10:38Z" level=info msg="ib-vip-sp21-21-07-390-9-2006-1-0-p9sbn is suspended, skipping execution" namespace=argo workflow=ib-vip-sp21-21-07-390-9-2006-1-0-p9sbn
time="2020-07-07T08:10:38Z" level=info msg="ib-vip-sp21-21-07-390-9-2006-1-0-f2nzp is suspended, skipping execution" namespace=argo workflow=ib-vip-sp21-21-07-390-9-2006-1-0-f2nzp
time="2020-07-07T08:10:38Z" level=info msg="ib-vip-sp21-21-07-390-9-2006-1-0-9ck8m is suspended, skipping execution" namespace=argo workflow=ib-vip-sp21-21-07-390-9-2006-1-0-9ck8m
time="2020-07-07T08:10:38Z" level=info msg="ib-vip-sp21-21-07-390-9-2006-1-0-lsh55 is suspended, skipping execution" namespace=argo workflow=ib-vip-sp21-21-07-390-9-2006-1-0-lsh55
time="2020-07-07T08:10:38Z" level=info msg="ib-vip-sp21-21-07-390-9-2006-1-0-22p29 is suspended, skipping execution" namespace=argo workflow=ib-vip-sp21-21-07-390-9-2006-1-0-22p29
time="2020-07-07T08:10:38Z" level=info msg="ib-vip-sp21-21-07-390-9-2006-1-0-m5skc is suspended, skipping execution" namespace=argo workflow=ib-vip-sp21-21-07-390-9-2006-1-0-m5skc
time="2020-07-07T08:10:41Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:41Z" level=info msg="Updating node exit-handler-bug-example-1275286746 status Pending -> Running"
time="2020-07-07T08:10:41Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:41Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:41Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=1405485 workflow=exit-handler-bug-example
time="2020-07-07T08:10:41Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:41Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:41Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:43Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:43Z" level=info msg="Setting node exit-handler-bug-example-2218368392 outputs"
time="2020-07-07T08:10:43Z" level=info msg="Updating node exit-handler-bug-example-2218368392 status Running -> Succeeded"
time="2020-07-07T08:10:43Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:43Z" level=info msg="Outbound nodes of exit-handler-bug-example-1359570492 set to [exit-handler-bug-example-2218368392]" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:43Z" level=info msg="node exit-handler-bug-example-1359570492 phase Running -> Succeeded" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:43Z" level=info msg="node exit-handler-bug-example-1359570492 finished: 2020-07-07 08:10:43.5745382 +0000 UTC" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:43Z" level=info msg="Checking daemoned children of exit-handler-bug-example-1359570492" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:43Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:43Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=1405493 workflow=exit-handler-bug-example
time="2020-07-07T08:10:43Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:43Z" level=info msg="Labeled pod argo/exit-handler-bug-example-2218368392 completed"
time="2020-07-07T08:10:43Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:43Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:50Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:50Z" level=info msg="Setting node exit-handler-bug-example-1275286746 outputs"
time="2020-07-07T08:10:50Z" level=info msg="Updating node exit-handler-bug-example-1275286746 status Running -> Succeeded"
time="2020-07-07T08:10:50Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:50Z" level=info msg="Outbound nodes of exit-handler-bug-example-2933386262 set to [exit-handler-bug-example-1275286746]" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:50Z" level=info msg="node exit-handler-bug-example-2933386262 phase Running -> Succeeded" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:50Z" level=info msg="node exit-handler-bug-example-2933386262 finished: 2020-07-07 08:10:50.8049955 +0000 UTC" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:50Z" level=info msg="Checking daemoned children of exit-handler-bug-example-2933386262" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:50Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:50Z" level=info msg="Outbound nodes of exit-handler-bug-example set to [exit-handler-bug-example-3038135764 exit-handler-bug-example-3071691002]" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:50Z" level=info msg="node exit-handler-bug-example phase Running -> Succeeded" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:50Z" level=info msg="node exit-handler-bug-example finished: 2020-07-07 08:10:50.8053449 +0000 UTC" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:50Z" level=info msg="Checking daemoned children of exit-handler-bug-example" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:50Z" level=info msg="Updated phase Running -> Succeeded" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:50Z" level=info msg="Marking workflow completed" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:50Z" level=info msg="Checking daemoned children of " namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:50Z" level=info msg="Workflow update successful" namespace=argo phase=Succeeded resourceVersion=1405511 workflow=exit-handler-bug-example
time="2020-07-07T08:10:50Z" level=info msg="Labeled pod argo/exit-handler-bug-example-1275286746 completed"
```

**Logs**

```
argo get <workflowname>
Name:                exit-handler-bug-example
Namespace:           argo
ServiceAccount:      default
Status:              Succeeded
Conditions:          
 Completed           True
Created:             Tue Jul 07 10:10:16 +0200 (10 minutes ago)
Started:             Tue Jul 07 10:10:16 +0200 (10 minutes ago)
Finished:            Tue Jul 07 10:10:50 +0200 (10 minutes ago)
Duration:            34 seconds
ResourcesDuration:   1m23s*(1 cpu),1m23s*(100Mi memory)

STEP                         TEMPLATE       PODNAME                              DURATION  MESSAGE
 ✔ exit-handler-bug-example  dag                                                             
 ├-✔ step-1                  step-template  exit-handler-bug-example-3071691002  10s         
 ├-✔ step-2                  step-template  exit-handler-bug-example-3054913383  8s          
 ├-✔ step-3                  step-template  exit-handler-bug-example-3038135764  8s          
 ├-✔ step-1.onExit           exit-handler                                                    
 | └-✔ exit-1                sleep-10       exit-handler-bug-example-2218368392  14s         
 └-✔ step-3.onExit           exit-handler                                                    
   └-✔ exit-1                sleep-10       exit-handler-bug-example-1275286746  13s

argo logs <workflowname>
exit-handler-bug-example-3054913383: step exit-handler-bug-example-3054913383
exit-handler-bug-example-3071691002: step exit-handler-bug-example-3071691002
exit-handler-bug-example-3038135764: step exit-handler-bug-example-3038135764
exit-handler-bug-example-2218368392: exit-handler-step-exit-handler-bug-example-2218368392
exit-handler-bug-example-1275286746: exit-handler-step-exit-handler-bug-example-1275286746

NOT APPLICABLE SINCE POD IS NEVER STARTED: kubectl logs <failedpodname> -c init

NOT APPLICABLE SINCE POD IS NEVER STARTED: kubectl logs <failedpodname> -c wait

kubectl logs -n argo $(kubectl get pods -l app=workflow-controller -n argo -o name)
time="2020-07-07T08:10:16Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:16Z" level=info msg="Updated phase  -> Running" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:16Z" level=info msg="DAG node exit-handler-bug-example initialized Running" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:16Z" level=info msg="All of node exit-handler-bug-example.step-1 dependencies [] completed" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:16Z" level=info msg="Pod node exit-handler-bug-example-3071691002 initialized Pending" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:16Z" level=info msg="Created pod: exit-handler-bug-example.step-1 (exit-handler-bug-example-3071691002)" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:16Z" level=info msg="All of node exit-handler-bug-example.step-2 dependencies [] completed" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:16Z" level=info msg="Pod node exit-handler-bug-example-3054913383 initialized Pending" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:16Z" level=info msg="Created pod: exit-handler-bug-example.step-2 (exit-handler-bug-example-3054913383)" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:16Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=1405366 workflow=exit-handler-bug-example
time="2020-07-07T08:10:16Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:16Z" level=info msg="Updating node exit-handler-bug-example-3071691002 message: ContainerCreating"
time="2020-07-07T08:10:16Z" level=info msg="Updating node exit-handler-bug-example-3054913383 message: ContainerCreating"
time="2020-07-07T08:10:17Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=1405373 workflow=exit-handler-bug-example
time="2020-07-07T08:10:17Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:17Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:21Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:21Z" level=info msg="Updating node exit-handler-bug-example-3054913383 status Pending -> Running"
time="2020-07-07T08:10:21Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=1405393 workflow=exit-handler-bug-example
time="2020-07-07T08:10:21Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:21Z" level=info msg="Updating node exit-handler-bug-example-3071691002 status Pending -> Running"
time="2020-07-07T08:10:22Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=1405394 workflow=exit-handler-bug-example
time="2020-07-07T08:10:22Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:22Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:26Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:26Z" level=info msg="Setting node exit-handler-bug-example-3054913383 outputs"
time="2020-07-07T08:10:26Z" level=info msg="Updating node exit-handler-bug-example-3054913383 status Running -> Succeeded"
time="2020-07-07T08:10:26Z" level=info msg="All of node exit-handler-bug-example.step-3 dependencies [step-2] completed" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:26Z" level=info msg="Pod node exit-handler-bug-example-3038135764 initialized Pending" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:26Z" level=info msg="Created pod: exit-handler-bug-example.step-3 (exit-handler-bug-example-3038135764)" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:26Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=1405411 workflow=exit-handler-bug-example
time="2020-07-07T08:10:26Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:26Z" level=info msg="Labeled pod argo/exit-handler-bug-example-3054913383 completed"
time="2020-07-07T08:10:26Z" level=info msg="Setting node exit-handler-bug-example-3071691002 outputs"
time="2020-07-07T08:10:26Z" level=info msg="Updating node exit-handler-bug-example-3038135764 message: ContainerCreating"
time="2020-07-07T08:10:27Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=1405418 workflow=exit-handler-bug-example
time="2020-07-07T08:10:27Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:27Z" level=info msg="Labeled pod argo/exit-handler-bug-example-3054913383 completed"
time="2020-07-07T08:10:27Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:28Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:28Z" level=info msg="Updating node exit-handler-bug-example-3071691002 status Running -> Succeeded"
time="2020-07-07T08:10:28Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:28Z" level=info msg="DAG node exit-handler-bug-example-1359570492 initialized Running" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:28Z" level=info msg="All of node step-1.onExit.exit-1 dependencies [] completed" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:28Z" level=info msg="Pod node exit-handler-bug-example-2218368392 initialized Pending" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:28Z" level=info msg="Created pod: step-1.onExit.exit-1 (exit-handler-bug-example-2218368392)" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:29Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=1405430 workflow=exit-handler-bug-example
time="2020-07-07T08:10:29Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:29Z" level=info msg="Labeled pod argo/exit-handler-bug-example-3071691002 completed"
time="2020-07-07T08:10:29Z" level=info msg="Updating node exit-handler-bug-example-2218368392 message: ContainerCreating"
time="2020-07-07T08:10:29Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:29Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=1405440 workflow=exit-handler-bug-example
time="2020-07-07T08:10:29Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:29Z" level=info msg="Labeled pod argo/exit-handler-bug-example-3071691002 completed"
time="2020-07-07T08:10:29Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:29Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:29Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:31Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:31Z" level=info msg="Updating node exit-handler-bug-example-3038135764 status Pending -> Running"
time="2020-07-07T08:10:31Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:31Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=1405447 workflow=exit-handler-bug-example
time="2020-07-07T08:10:31Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:31Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:33Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:33Z" level=info msg="Updating node exit-handler-bug-example-2218368392 status Pending -> Running"
time="2020-07-07T08:10:33Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:34Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=1405453 workflow=exit-handler-bug-example
time="2020-07-07T08:10:34Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:34Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:36Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:36Z" level=info msg="Setting node exit-handler-bug-example-3038135764 outputs"
time="2020-07-07T08:10:36Z" level=info msg="Updating node exit-handler-bug-example-3038135764 status Running -> Succeeded"
time="2020-07-07T08:10:36Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:36Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:36Z" level=info msg="DAG node exit-handler-bug-example-2933386262 initialized Running" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:36Z" level=info msg="All of node step-3.onExit.exit-1 dependencies [] completed" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:36Z" level=info msg="Pod node exit-handler-bug-example-1275286746 initialized Pending" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:36Z" level=info msg="Created pod: step-3.onExit.exit-1 (exit-handler-bug-example-1275286746)" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:36Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=1405466 workflow=exit-handler-bug-example
time="2020-07-07T08:10:36Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:36Z" level=info msg="Labeled pod argo/exit-handler-bug-example-3038135764 completed"
time="2020-07-07T08:10:36Z" level=info msg="Updating node exit-handler-bug-example-1275286746 message: ContainerCreating"
time="2020-07-07T08:10:36Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:36Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:37Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=1405474 workflow=exit-handler-bug-example
time="2020-07-07T08:10:37Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:37Z" level=info msg="Labeled pod argo/exit-handler-bug-example-3038135764 completed"
time="2020-07-07T08:10:37Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:37Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:37Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:37Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:37Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:37Z" level=info msg="ib-vip-sp21-21-07-390-9-2006-1-0-m7hk7 is suspended, skipping execution" namespace=argo workflow=ib-vip-sp21-21-07-390-9-2006-1-0-m7hk7
time="2020-07-07T08:10:38Z" level=info msg="ib-vip-sp21-21-07-390-9-2006-1-0-htf86 is suspended, skipping execution" namespace=argo workflow=ib-vip-sp21-21-07-390-9-2006-1-0-htf86
time="2020-07-07T08:10:38Z" level=info msg="ib-vip-sp21-21-07-390-9-2006-1-0-p9sbn is suspended, skipping execution" namespace=argo workflow=ib-vip-sp21-21-07-390-9-2006-1-0-p9sbn
time="2020-07-07T08:10:38Z" level=info msg="ib-vip-sp21-21-07-390-9-2006-1-0-f2nzp is suspended, skipping execution" namespace=argo workflow=ib-vip-sp21-21-07-390-9-2006-1-0-f2nzp
time="2020-07-07T08:10:38Z" level=info msg="ib-vip-sp21-21-07-390-9-2006-1-0-9ck8m is suspended, skipping execution" namespace=argo workflow=ib-vip-sp21-21-07-390-9-2006-1-0-9ck8m
time="2020-07-07T08:10:38Z" level=info msg="ib-vip-sp21-21-07-390-9-2006-1-0-lsh55 is suspended, skipping execution" namespace=argo workflow=ib-vip-sp21-21-07-390-9-2006-1-0-lsh55
time="2020-07-07T08:10:38Z" level=info msg="ib-vip-sp21-21-07-390-9-2006-1-0-22p29 is suspended, skipping execution" namespace=argo workflow=ib-vip-sp21-21-07-390-9-2006-1-0-22p29
time="2020-07-07T08:10:38Z" level=info msg="ib-vip-sp21-21-07-390-9-2006-1-0-m5skc is suspended, skipping execution" namespace=argo workflow=ib-vip-sp21-21-07-390-9-2006-1-0-m5skc
time="2020-07-07T08:10:41Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:41Z" level=info msg="Updating node exit-handler-bug-example-1275286746 status Pending -> Running"
time="2020-07-07T08:10:41Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:41Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:41Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=1405485 workflow=exit-handler-bug-example
time="2020-07-07T08:10:41Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:41Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:41Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:43Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:43Z" level=info msg="Setting node exit-handler-bug-example-2218368392 outputs"
time="2020-07-07T08:10:43Z" level=info msg="Updating node exit-handler-bug-example-2218368392 status Running -> Succeeded"
time="2020-07-07T08:10:43Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:43Z" level=info msg="Outbound nodes of exit-handler-bug-example-1359570492 set to [exit-handler-bug-example-2218368392]" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:43Z" level=info msg="node exit-handler-bug-example-1359570492 phase Running -> Succeeded" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:43Z" level=info msg="node exit-handler-bug-example-1359570492 finished: 2020-07-07 08:10:43.5745382 +0000 UTC" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:43Z" level=info msg="Checking daemoned children of exit-handler-bug-example-1359570492" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:43Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:43Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=1405493 workflow=exit-handler-bug-example
time="2020-07-07T08:10:43Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:43Z" level=info msg="Labeled pod argo/exit-handler-bug-example-2218368392 completed"
time="2020-07-07T08:10:43Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:43Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:50Z" level=info msg="Processing workflow" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:50Z" level=info msg="Setting node exit-handler-bug-example-1275286746 outputs"
time="2020-07-07T08:10:50Z" level=info msg="Updating node exit-handler-bug-example-1275286746 status Running -> Succeeded"
time="2020-07-07T08:10:50Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:50Z" level=info msg="Outbound nodes of exit-handler-bug-example-2933386262 set to [exit-handler-bug-example-1275286746]" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:50Z" level=info msg="node exit-handler-bug-example-2933386262 phase Running -> Succeeded" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:50Z" level=info msg="node exit-handler-bug-example-2933386262 finished: 2020-07-07 08:10:50.8049955 +0000 UTC" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:50Z" level=info msg="Checking daemoned children of exit-handler-bug-example-2933386262" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:50Z" level=info msg="Running OnExit handler: exit-handler" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:50Z" level=info msg="Outbound nodes of exit-handler-bug-example set to [exit-handler-bug-example-3038135764 exit-handler-bug-example-3071691002]" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:50Z" level=info msg="node exit-handler-bug-example phase Running -> Succeeded" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:50Z" level=info msg="node exit-handler-bug-example finished: 2020-07-07 08:10:50.8053449 +0000 UTC" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:50Z" level=info msg="Checking daemoned children of exit-handler-bug-example" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:50Z" level=info msg="Updated phase Running -> Succeeded" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:50Z" level=info msg="Marking workflow completed" namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:50Z" level=info msg="Checking daemoned children of " namespace=argo workflow=exit-handler-bug-example
time="2020-07-07T08:10:50Z" level=info msg="Workflow update successful" namespace=argo phase=Succeeded resourceVersion=1405511 workflow=exit-handler-bug-example
time="2020-07-07T08:10:50Z" level=info msg="Labeled pod argo/exit-handler-bug-example-1275286746 completed"
```

---

**Message from the maintainers**:

If you are impacted by this bug please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

**Comment by alexec** [comment]
@simster7 can you please investigate?
